A partial rewrite of java.vim

agaman opened the pull request:
Works the same way as before, using cindent() for most indenting and handling Java-specific cases.

Handles some edge cases better, like methods which have annotations and "throws" statements.

Improves readability by breaking out reusable bits like line-matching regexes into individual functions.

Commit:
A partial rewrite of java.vim

Works the same way as before, using cindent() for most indenting and handling Java-specific cases.
Handles some edge cases better, like methods which have annotations and "throws" statements.
Improves readability by breaking out reusable bits like line-matching regexes into individual functions.

xuhdev commented:
Cool, thanks! But I see some indent style change, while you added `vi: sw=2 et`, you use 4 space indent style. Could you fix this please?

You could make the change, then `git commit -a --amend`, `git push origin your_branch --force`, in this case your branch is `master`.

I'll do more tests tonight (GMT -7).

agaman commented:
Hey Hong, the indentation has been updated to 2 spaces. Sorry about that,
I'm used to developing with 4 space indents.

Alex Gaman
agaman@gmail.com

On Thu, Apr 18, 2013 at 2:08 PM, Hong Xu notifications@github.com wrote:

> Cool, thanks! But I see some indent style change, while you added vi:
> sw=2 et, you use 4 space indent style. Could you fix this please?
> 
> You could make the change, then git commit -a --amend, git push origin
> your_branch --force, in this case your branch is master.
> 
> I'll do more tests tonight (GMT -7).
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/xuhdev/indent-java.vim/pull/1#issuecomment-16593285
> .

xuhdev commented on indent/java.vim:
```diff
@@ -19,129 +19,183 @@ let b:did_indent = 1
 
 " Set the function to do the work.
-setlocal indentexpr=GetJavaIndent()
+setlocal indentexpr=JavaIndent()
```
I can't see the point of renaming this function. Do you have any good reasons?

agaman replied:
Mostly keeping in line with the vim function cindent(). I figure if we are
indenting C code with cindent(), we should indent Java code with
JavaIndent().

Alex Gaman
agaman@gmail.com

On Thu, Apr 18, 2013 at 2:40 PM, Hong Xu notifications@github.com wrote:

> In indent/java.vim:
> 
> >  " Set the function to do the work.
> > -setlocal indentexpr=GetJavaIndent()
> > +setlocal indentexpr=JavaIndent()
> 
> I can't see the point of renaming this function. Do you have any good
> reasons?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/xuhdev/indent-java.vim/pull/1/files#r3860351
> .

xuhdev replied:
Thanks for your reply. Actually I inspected [the list of indent files](http://code.google.com/p/vim/source/browse/#hg%2Fruntime%2Findent), and it turns out that most use the `GetXxxIndent` style. I think it's just a vimscript style. Could you get it back to `GetJavaIndent()` please?

agaman replied:
Aha, point taken. I updated the commit to reflect that naming scheme.

Alex Gaman
agaman@gmail.com

On Thu, Apr 18, 2013 at 9:38 PM, Hong Xu notifications@github.com wrote:

> In indent/java.vim:
> 
> >  " Set the function to do the work.
> > -setlocal indentexpr=GetJavaIndent()
> > +setlocal indentexpr=JavaIndent()
> 
> Thanks for your reply. Actually I inspected the list of indent fileshttp://code.google.com/p/vim/source/browse/#hg%2Fruntime%2Findent,
> and it turns out that most use the GetXxxIndent style. I think it's just
> a vimscript style. Could you get it back to GetJavaIndent() please?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/xuhdev/indent-java.vim/pull/1/files#r3866611
> .

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
   " specific cases.
-  let theIndent = cindent(v:lnum)
+  let defaultIndent = cindent(currentLineNum)
+  echo "defaultIndent " defaultIndent
```
Could you explain the reason of having an echo here? Thanks

agaman replied:
That was there for debugging. Definitely should have been removed before committing. Apologies

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+endfunction
+
+function! IsAnnotation(text)
+  return a:text =~ '^\s*@'
```
I noticed that you shortened the previous `'^\s*@\S\+\s*$'` to `'^\s*@'`. It's not a much large problem, but I would say that the original one may be better at handling this case:

``` java
/* This comment is written by
@someone */
```

Your regex is not able to handle this case.

agaman replied:
Hrm, that's true, but it does account for most of the use cases of an annotation. Your case of an annotation within a comment is actually part of a larger case of "is the given line within a comment block," and I don't think we can solve that with regexes. I will look into making a better IsComment function which takes into account the line's context.

From Wikipedia, I have the following examples of fairly complex annotations which would not get caught by the previous, stricter regex:

``` java
@UnderConstruction(owner="Navin Gujarish")
```

and:

``` java
@Target({ElementType.TYPE,ElementType.METHOD, ElementType.CONSTRUCTOR,ElementType.ANNOTATION_TYPE, ElementType.PACKAGE,ElementType.FIELD,ElementType.LOCAL_VARIABLE})
```

xuhdev replied:
Yes you're right. Yours should be better.

xuhdev replied:
What about:

``` viml
return a:text =~ '^\s*@' && !IsComment(a:text)
```

?

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
-        \ && getline(lnum) !~ '^\s*\(extends\|implements\)\>'
-    let theIndent = theIndent + &sw
+  " Annotations should not be more indented than the previous comment line
+  if IsAnnotation(currentLineText) && IsComment(prevLineText)
```
Here you have `IsComment(prevLineText)`, I would say it is always `false`. Since the `prevLineText` you got here is the previous non-blank and non-comment line.

agaman replied:
Good point. I will look into what my use case was here.

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+  if IsMethodDetail(currentLineText)
+    if IsMethodDetail(prevLineText)
+      return indent(prevLineNum)
+    else
```
Could you remove this `else` clause please? It isn't necessary here and makes the code a little bit wordy.

xuhdev replied:
Not exactly "wordy" -- it's a matter of structure

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+      " We want "throws" to be indented more than function defintion
+      " lines but less than their arguments
+      let prevIndent = indent(prevLineNum)
+      if prevIndent <= defaultIndent
```
Maybe it should be `prevIndent < defaultIndent` without the `=` case?

Consider the following case:

``` java
public test(
        int a,
        int b
        )
// Current line
```

On the current line, `defaultIndent` should equal to the previous line. But you want `prevIndent - &sw` right?

Also, could you please add one detail in the comments?

> ... less than their arguments when the arguments are listed in multiple lines

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+  if IsClassDetail(currentLineText)
+    if IsClassDetail(prevLineText)
+      return indent(prevLineNum)
+    else
```
Same `else` here

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+      " We want "implements" and "extends" to be indented
+      " more than class defintion lines
+      let prevIndent = indent(prevLineNum)
+      if prevIndent <= defaultIndent
```
Could you explain the point here, please? I can understand the comparison between `prevIndent` and `defaultIndent` in the `throws` part, but I cannot understand the point here.

agaman replied:
I found that the defaultIndent for "implements" and "extends" was on the same indentation as the class definition in some cases. For instance, it was:

``` java
class MyClass
implements MyInterface
{
}
```

Instead of

``` java
class MyClass
    implements MyInterface
{
}
```

The "else" in this case is absolutely unnecessary and just an artifact of reusing the "throws" code, however.

xuhdev replied:
OK, so maybe removing the "else" clause could make the code cleaner.

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+endfunction
+
+function! IsMethodDetail(text)
+  return a:text =~ '^\s*throws\>'
```
In fact the regex is not correct (I know you get this from the previous one). Could you fix this please? Because lines like

`throws_me = 12`

could show up in Java code, and `iskeyword` could be defined by users to eliminate `_` as a keyword.

I personally think `^\s*throws\s\+` might be better.

Same story for `IsClassDetail`

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
-        let theIndent = 0
+  " Align "throws" lines for methods
+  if IsMethodDetail(currentLineText)
+    if IsMethodDetail(prevLineText)
```
Could you give an example of two consecutive `throws` lines?

agaman replied:
Good point. This code should look for multiline "throw" statements instead like:

``` java
public void doSomething()
    throws  AnException,
            AnotherException
            YetAnotherException
{
}

```

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+  if IsMethodDetail(prevLineText)
+    if EndsInComma(prevLineText)
+      return defaultIndent
+    else
```
Add

``` viml
if prevLineText =~ '^\s*throws\s*$'
```

to handle the single "throws" case.

Also, it seems that the following case is not handled well:

``` java
public fun()
    throws Exception {
// current line
```

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+endfunction
+
+function! IsOpenBrace(text)
+  return a:text =~ '{[^}]*$'
```
For example, `{{}` is not handled properly for `IsOpenBrace`function.

xuhdev replied:
Seems that this function is actually not used

agaman replied:
True. I think this was added when I was trying to figure out how to match braces.

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+endfunction
+
+function! IsCloseBrace(text)
+  return a:text =~ '^\s*}\s*\(//.*\|/\*.*\)\=$'
```
Seems this doesn't handle every case well.

For example anonymous class:

``` java
anything.addListener(new SomeAdapter() {
    //...
});
```

agaman replied:
Good point.

Do you think a regex like this would be more appropriate?:

``` perl
'^\s*[})\]]'
```

This way we indent properly for the first instance of some kind of closing brace (any } or ) or ]), and it doesn't matter what comes after the brace. eg:

``` java
anything.addListener(new SomeAdapter() {
    //...
}).addListener(otherAdapter);
```

xuhdev replied:
I agree with you on this.

xuhdev commented on indent/java.vim:
```diff
@@ -35,113 +35,167 @@ endif
+  while currentLineNum > 1 && (  IsMethodDetail(currentLineText)
+    || IsClassDetail(currentLineText)
+    || EndsInComma(currentLineText)
+    )
```
Should be 

``` viml
   while currentLineNum > 1 && (  IsMethodDetail(currentLineText)
     \ || IsClassDetail(currentLineText)
     \ || EndsInComma(currentLineText) )
```

Or there is a syntax error

agaman replied:
Weird, I will double-check that this is doing what I think it's doing. Definitely will add the backslashes.

xuhdev commented:
Thanks for your huge patch, I really appreciate it, since it's a lot of work. The comments above are the problems that I found, please let me know if you find any of them inappropriate.

agaman commented:
Thanks for the feedback Hong, there is definitely more work that needs to be done to clean up this code and I am doing my best to make it as readable as possible and accurate as possible without rewriting the entirety of the indentation rules. I recently taught myself vimscript just so that I could update this indentation code. The feedback on syntax errors and needless verbosity is greatly appreciated.

I will have another commit sometime in the next couple days with the above issues addressed as well as some more fixes/algorithm changes (including a function which identifies comments better).

Commit:
Added a bunch of functions which makes indentation aware of multiple line statements

Commit:
Fixed an issue where multi-line function definitions with a closing paren on the last line messed up indentation after a "throws" statement.

Commit:
Fixing bug in "IsComment"

It couldn't distinguish single-line block comments from open block comment lines
Also fixed an erroneous deletion and some code comments

Commit:
Fixing issue with class annotations.

Fixed a recently introduced bug with detecting comments
Made sure open braces after classes are aligned correctly
Fixed an issue with throws statements on the same line as open braces

xuhdev commented:
Hi, Alex, have you finished all the modifications?

agaman commented:
Hey Hong, I actually have one more change I will be committing in the next couple days.

Commit:
Merge branch 'master' of github.com:xuhdev/indent-java.vim

Commit:
Adding test cases and fixing code to work with tests

agaman commented:
For some reason my latest merge and commit don't have my name on them. It may have something to do with my squashing commits on another branch. That said, I assure you that this latest commit is from me.

xuhdev commented:
Thanks for your responding. I can merge manually with `--squash` option to merge your commits as a single commit, with your name and email as the author.
